### Python Fundamentals for Data Analysts

# 4.5: Data Consistency Checks

### Content

1. Perform data consistency on Products dataframe
- Check missing data: df_prods.isnull().sum()
- Check mixed data type
- Check duplicates: df_prods_clean[df_prods_clean.duplicated()]

2. Perform data consistency on Order dataframe
- Check missing data
- Check mixed data type
- Check duplicates

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import os

#### Load data files

In [2]:
# Set path to access data files & Load data file
path = r'C:\Users\tsoew\OneDrive\Desktop\InstaCart Basket Analysis'

# Loading products.csv
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)
# Loading orders.csv
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

### #1 Perform Data Consistency on Products Data

In [3]:
# Check df_prods content
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [4]:
# Check descriptive statistic
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### Step 1: Check missing data

In [5]:
# Checking for missing data in products dataframe
# There are 16 missing data in product_name column
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# Print the rows with missing values
df_prods[df_prods['product_name'].isnull()]

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
# Getting a new dataframe without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [8]:
df_prods.shape

(49693, 5)

In [9]:
df_prods_clean.shape

(49677, 5)

**Note:**  The reduced number of 16 rows show that the missing rows have been removed.

#### Step 2: Check Mixed Data

In [10]:
# Checking if there is any mix data type in a column

# run loop for every column in df_test
for col in df_prods.columns.tolist(): 
    
    # variable weird is to be used to determine there is mixed data type
    weird = (df_prods[[col]].map(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
    
    if len (df_prods[weird]) > 0:
        print (col)

product_name


**Note:** product_name column has mix data

In [11]:
# Check the mixed type again with the new cleaned dataframe

# run loop for every column in df_test
for col in df_prods_clean.columns.tolist(): 
    
    # variable weird is to be used to determine there is mixed data type
    weird = (df_prods_clean[[col]].map(type) != df_prods_clean[[col]].iloc[0].apply(type)).any(axis = 1)
    
    if len (df_prods_clean[weird]) > 0:
        print (col)

**Note:** the mix data is gone. That means it comes from the null values in product_name column

#### Step 3: Check Duplicates

In [12]:
# Checking for any duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

**Note:** Result: there are 5 duplicate data

In [14]:
# Drop the duplicates and assign it to a new variable
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
# The new variable has 5 less rows, which means the duplicates have been removed.
df_prods_clean_no_dups.shape

(49672, 5)

### #2 Perform consistency test on Orders Data

In [16]:
# Check content
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [17]:
# Run df.describe() function to get descriptive statistics
# Result: all minimum and maximum values seem to be normal
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Step 1: Check Mixed Data

In [18]:
# Check for mixed-type data in your df_ords dataframe.

# run loop for every column in df_test
for col in df_ords.columns.tolist(): 
    
    # variable weird is to be used to determine there is mixed data type
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    
    if len (df_ords[weird]) > 0:
        print (col)

**Note:** Result: there is no mixed data in this Dataframe.

#### Step 2: Check Missing Data

In [19]:
# Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

**Note**: There are 206209 missing values in days_since_prior_order column

In [20]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [21]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()]

In [22]:
#Display the missing values
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#### Explanation of missing values: 

It seems the missing values occur everytime a new user_id is entered. This makes sense because as a first time user, there is no prior order; therefore NaN value. We will leave it the way it is because it will not affect any operation.

#### Step 3: Check for Duplicates

In [23]:
# Checking for any duplicates
# Result: there are NO duplicate data
df_ords_dups = df_ords[df_ords.duplicated()]

In [24]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

**Note:** There is no duplicate on this data

### Export Data back to files

In [25]:
# Writing cleaned product datafame to products_clean.csv
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_clean.csv'), index=False)

# Writing cleaned order dataframe to orders_clean.csv
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_clean.csv'), index=False)

# End